src/platform/qt/library/LibraryModel.cpp

```diff
@@ -0,0 +1,422 @@
+/* Copyright (c) 2014-2017 waddlesplash
+ * Copyright (c) 2013-2022 Jeffrey Pfau
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include "LibraryModel.h"
+
+#include "utils.h"
+
+#include <QSortFilterProxyModel>
+#include <QItemSelectionModel>
+#include <QApplication>
+#include <QDir>
+#include <algorithm>
+
+using namespace QGBA;
+
+namespace QGBA {
+
+struct SpanSet {
+	struct Span {
+		int left;
+		int right;
+
+		bool operator<(const Span& other) const { return left < other.left; }
+		bool operator>(const Span& other) const { return left > other.left; }
+	};
+
+	void add(int pos) {
+		bool found = false;
+		for (Span& span : spans) {
+			if (pos == span.left - 1) {
+				span.left = pos;
+				found = true;
+			} else if (pos == span.right + 1) {
+				span.right = pos;
+				found = true;
+			}
+		}
+		if (!found) {
+			spans << Span{ pos, pos };
+			return;
+		}
+	}
+
+	void merge() {
+		int numSpans = spans.size();
+		for (int i = 0; i < numSpans; i++) {
+			int aLeft = spans[i].left - 1;
+			int aRight = spans[i].right + 1;
+			for (int j = numSpans - 1; j > i; j--) {
+				int bLeft = spans[j].left - 1;
+				int bRight = spans[j].right + 1;
+				if (aLeft < bRight && aRight > bLeft) {
+					aLeft = aLeft < bLeft ? aLeft : bLeft;
+					aRight = aRight > bRight ? aRight : bRight;
+					spans[i].left = aLeft;
+					spans[i].right = aRight;
+					spans.removeAt(j);
+					--numSpans;
+				}
+			}
+		}
+	}
+
+	void sort(bool reverse = false) {
+		if (reverse) {
+			std::sort(spans.begin(), spans.end(), std::greater<Span>());
+		} else {
+			std::sort(spans.begin(), spans.end());
+		}
+	}
+
+	QList<Span> spans;
+};
+
+}
+
+LibraryModel::LibraryModel(LibraryController* parent)
+	: QAbstractItemModel(parent)
+	, m_controller(parent)
+	, m_treeMode(false)
+	, m_showFilename(false)
+{
+}
+
+bool LibraryModel::treeMode() const
+{
+	return m_treeMode;
+}
+
+void LibraryModel::setTreeMode(bool tree)
+{
+	if (m_treeMode == tree) {
+		return;
+	}
+	beginResetModel();
+	m_treeMode = tree;
+	endResetModel();
+}
+
+bool LibraryModel::showFilename() const
```

Comment:
Nit: No newline after the function signature (constructors are an exception).

Reply:
Oops. I _knew_ that was the style guide, but I follow the Linux kernel style guide in basically everything else so I forgot ^^()